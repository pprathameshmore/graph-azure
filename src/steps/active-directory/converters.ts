import map from 'lodash.map';

import {
  convertProperties,
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  getTime,
  IntegrationInstance,
  Relationship,
  RelationshipDirection,
  assignTags,
  createMappedRelationship,
  setRawData,
} from '@jupiterone/integration-sdk-core';
import { Group, Organization, User } from '@microsoft/microsoft-graph-types';

import {
  generateEntityKey,
  generateRelationshipKey,
} from '../../utils/generateKeys';
import {
  GroupMember,
  MemberType,
  IdentitySecurityDefaultsEnforcementPolicy,
} from './client';
import {
  ACCOUNT_ENTITY_CLASS,
  ACCOUNT_ENTITY_TYPE,
  ACCOUNT_GROUP_RELATIONSHIP_TYPE,
  GROUP_ENTITY_CLASS,
  GROUP_ENTITY_TYPE,
  GROUP_MEMBER_ENTITY_CLASS,
  GROUP_MEMBER_ENTITY_TYPE,
  GROUP_MEMBER_RELATIONSHIP_TYPE,
  USER_ENTITY_CLASS,
  USER_ENTITY_TYPE,
  SERVICE_PRINCIPAL_ENTITY_CLASS,
  SERVICE_PRINCIPAL_ENTITY_TYPE,
} from './constants';
import { RelationshipClass } from '@jupiterone/integration-sdk-core';

export function createAccountEntity(instance: IntegrationInstance): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {},
      assign: {
        _class: ACCOUNT_ENTITY_CLASS,
        _key: generateEntityKey(instance.id),
        _type: ACCOUNT_ENTITY_TYPE,
        name: instance.name,
        displayName: instance.name,
      },
    },
  });
}

export function createAccountEntityWithOrganization(
  instance: IntegrationInstance,
  organization: Organization,
  securityDefaults?: IdentitySecurityDefaultsEnforcementPolicy,
): Entity {
  let defaultDomain: string | undefined;
  const verifiedDomains = map(organization.verifiedDomains, (e) => {
    if (e.isDefault) {
      defaultDomain = e.name;
    }
    return e.name as string;
  });

  const accountEntityWithOrganization = createIntegrationEntity({
    entityData: {
      source: organization,
      assign: {
        _class: ACCOUNT_ENTITY_CLASS,
        _key: generateEntityKey(instance.id),
        _type: ACCOUNT_ENTITY_TYPE,
        name: organization.displayName,
        displayName: instance.name,
        organizationName: organization.displayName,
        defaultDomain,
        verifiedDomains,
        securityDefaultsEnabled: securityDefaults?.isEnabled,
      },
    },
  });

  if (securityDefaults) {
    setRawData(accountEntityWithOrganization, {
      name: 'identitySecurityDefaultsEnforcementPolicy',
      rawData: securityDefaults,
    });
  }
  return accountEntityWithOrganization;
}

export function createGroupEntity(data: Group): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data, { parseTime: true }),
        _key: generateEntityKey(data.id),
        _class: GROUP_ENTITY_CLASS,
        _type: GROUP_ENTITY_TYPE,
        name: data.displayName,
        deletedOn: getTime(data.deletedDateTime),
        createdOn: getTime(data.createdDateTime),
        email: data.mail ?? undefined,
        renewedOn: getTime(data.renewedDateTime),
      },
    },
  });
}

export function createUserEntity(data: User): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: generateEntityKey(data.id),
        _class: USER_ENTITY_CLASS,
        _type: USER_ENTITY_TYPE,
        name: data.displayName,
        email: data.mail ?? undefined,
        firstName: data.givenName,
        lastName: data.surname,
        username: data.userPrincipalName,
      },
    },
  });
}

export function createServicePrincipalEntity(data: any): Entity {
  const entity = createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(data.id),
        _class: SERVICE_PRINCIPAL_ENTITY_CLASS,
        _type: SERVICE_PRINCIPAL_ENTITY_TYPE,
        userType: 'service',
        category: ['infrastructure'],
        name: data.displayName,
        displayName: data.displayName,
        appDisplayName: data.appDisplayName,
        appId: data.appId,
        servicePrincipalType: data.servicePrincipalType,
        servicePrincipalNames: data.servicePrincipalNames,
      },
    },
  });

  assignTags(entity, data.tags);
  return entity;
}

export function createAccountGroupRelationship(
  account: Entity,
  group: Entity,
): Relationship {
  const parentKey = account._key;
  const childKey = generateEntityKey(group.id);

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromKey: parentKey,
    fromType: ACCOUNT_ENTITY_TYPE,
    toKey: childKey,
    toType: GROUP_ENTITY_TYPE,
    properties: {
      _type: ACCOUNT_GROUP_RELATIONSHIP_TYPE,
    },
  });
}

export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  const fromKey = account._key;
  const toKey = generateEntityKey(user.id);

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromType: ACCOUNT_ENTITY_TYPE,
    fromKey,
    toType: USER_ENTITY_TYPE,
    toKey,
  });
}

export function createGroupMemberRelationship(
  group: Entity,
  member: GroupMember,
): Relationship {
  const memberEntityType = getGroupMemberEntityType(member);
  const memberEntityClass = getGroupMemberEntityClass(member);

  const groupKey = generateEntityKey(group.id);
  const memberKey = generateEntityKey(member.id);

  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _key: generateRelationshipKey(groupKey, memberKey),
    _type: GROUP_MEMBER_RELATIONSHIP_TYPE,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: groupKey,
      targetFilterKeys: [['_type', '_key']],
      targetEntity: {
        _type: memberEntityType,
        _class: memberEntityClass,
        _key: memberKey,
        displayName: member.displayName,
        jobTitle: member.jobTitle,
        email: member.mail,
      },
    },
    properties: {
      groupId: group.id as string,
      memberId: member.id,
      memberType: member['@odata.type'],
    },
  });
}

function getGroupMemberEntityType(member: GroupMember): string {
  switch (member['@odata.type']) {
    case MemberType.USER:
      return USER_ENTITY_TYPE;
    case MemberType.GROUP:
      return GROUP_ENTITY_TYPE;
    default:
      return GROUP_MEMBER_ENTITY_TYPE;
  }
}

function getGroupMemberEntityClass(member: GroupMember): string {
  switch (member['@odata.type']) {
    case MemberType.USER:
      return USER_ENTITY_CLASS;
    case MemberType.GROUP:
      return GROUP_ENTITY_CLASS;
    default:
      return GROUP_MEMBER_ENTITY_CLASS;
  }
}
