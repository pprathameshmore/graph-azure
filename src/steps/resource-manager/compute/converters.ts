import {
  DataDisk,
  Disk,
  Image,
  OSDisk,
  VirtualMachine,
  VirtualMachineExtension,
} from '@azure/arm-compute/esm/models';
import {
  assignTags,
  convertProperties,
  createIntegrationEntity,
  Entity,
  getTime,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { resourceGroupName } from '../../../azure/utils';
import {
  DISK_ENTITY_CLASS,
  DISK_ENTITY_TYPE,
  entities,
  VIRTUAL_MACHINE_ENTITY_CLASS,
  VIRTUAL_MACHINE_ENTITY_TYPE,
  VIRTUAL_MACHINE_IMAGE_ENTITY_CLASS,
  VIRTUAL_MACHINE_IMAGE_ENTITY_TYPE,
} from './constants';

export function createVirtualMachineEntity(
  webLinker: AzureWebLinker,
  data: VirtualMachine,
): Entity {
  const osProperties = {};
  if (data.storageProfile) {
    Object.assign(osProperties, {
      platform: data.storageProfile.osDisk?.osType?.toLowerCase(),
      osName: data.storageProfile.imageReference?.offer,
      osVersion: data.storageProfile.imageReference?.exactVersion,
    });
  }
  if (data.osProfile) {
    Object.assign(osProperties, {
      adminUser: data.osProfile.adminUsername,
      disablePasswordAuthentication:
        data.osProfile.linuxConfiguration?.disablePasswordAuthentication,
      enableAutomaticUpdates:
        data.osProfile.windowsConfiguration?.enableAutomaticUpdates,
      timeZone: data.osProfile.windowsConfiguration?.timeZone,
    });
  }
  const entity = {
    ...convertProperties(data),
    ...osProperties,
    _key: data.id as string,
    _type: VIRTUAL_MACHINE_ENTITY_TYPE,
    _class: VIRTUAL_MACHINE_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    vmId: data.vmId,
    type: data.type,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    state: data.provisioningState,
    vmSize: data.hardwareProfile && data.hardwareProfile.vmSize,
    usesManagedDisks: usesManagedDisks(
      data.storageProfile?.osDisk,
      data.storageProfile?.dataDisks,
    ),
    webLink: webLinker.portalResourceUrl(data.id),
  };

  assignTags(entity, data.tags);

  return entity;
}

function usesManagedDisks(osDisk?: OSDisk, dataDisks?: DataDisk[]) {
  function usesManagedDisk(disk: OSDisk | DataDisk | undefined) {
    return disk?.managedDisk !== undefined;
  }

  return (
    usesManagedDisk(osDisk) &&
    (dataDisks || []).every((d) => usesManagedDisk(d))
  );
}

export function createDiskEntity(
  webLinker: AzureWebLinker,
  data: Disk,
): Entity {
  const entity = {
    ...convertProperties(data),
    _key: data.id as string,
    _type: DISK_ENTITY_TYPE,
    _class: DISK_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    createdOn: getTime(data.timeCreated),
    webLink: webLinker.portalResourceUrl(data.id),
    encrypted: !!data.encryption?.type,
    encryption: data.encryption?.type,
    state: data.diskState?.toLowerCase(),
    attached: data.diskState === 'Attached',
  };

  assignTags(entity, data.tags);

  return entity;
}

export function createImageEntity(
  webLinker: AzureWebLinker,
  data: Image,
): Entity {
  const entity = {
    ...convertProperties(data),
    _key: data.id as string,
    _type: VIRTUAL_MACHINE_IMAGE_ENTITY_TYPE,
    _class: VIRTUAL_MACHINE_IMAGE_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    webLink: webLinker.portalResourceUrl(data.id),
  };

  assignTags(entity, data.tags);

  return entity;
}

export type VirtualMachineExtensionSharedProperties = Omit<
  VirtualMachineExtension,
  'id' | 'location' | 'provisioningState'
>;

export function createVirtualMachineExtensionEntity(
  data: VirtualMachineExtensionSharedProperties,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getVirtualMachineExtensionKey(data),
        _type: entities.VIRTUAL_MACHINE_EXTENSION._type,
        _class: entities.VIRTUAL_MACHINE_EXTENSION._class,
        name: data.name,
      },
    },
  });
}

export function getVirtualMachineExtensionKey(
  data: VirtualMachineExtensionSharedProperties,
) {
  return `vm-extension:${data.publisher || 'unknown-publisher'}:${data.name!}`;
}

export const testFunctions = {
  usesManagedDisks,
};
