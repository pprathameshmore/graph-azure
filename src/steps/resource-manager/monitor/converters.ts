import { AzureWebLinker } from '../../../azure';
import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';
import { MonitorEntities } from './constants';
import {
  DiagnosticSettingsResource,
  LogProfileResource,
  LogSettings,
} from '@azure/arm-monitor/esm/models';

export function createMonitorLogProfileEntity(
  webLinker: AzureWebLinker,
  data: LogProfileResource,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id,
        _type: MonitorEntities.MONITOR_LOG_PROFILE._type,
        _class: MonitorEntities.MONITOR_LOG_PROFILE._class,
        id: data.id,
        webLink: webLinker.portalResourceUrl(data.id),
        name: data.name,
        displayName: data.name,
        storageAccountId: data.storageAccountId,
        serviceBusRuleId: data.serviceBusRuleId,
        locations: data.locations,
        categories: data.categories,
        'retentionPolicy.enabled': data.retentionPolicy.enabled,
        'retentionPolicy.days': data.retentionPolicy.days,
      },
    },
  });
}

export function createDiagnosticSettingsEntity(
  webLinker: AzureWebLinker,
  data: DiagnosticSettingsResource,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id!,
        _type: MonitorEntities.DIAGNOSTIC_SETTINGS._type,
        _class: MonitorEntities.DIAGNOSTIC_SETTINGS._class,
        id: data.id,
        webLink: webLinker.portalResourceUrl(data.id),
        name: data.name,
        storageAccountId: data.storageAccountId,
        eventHubName: data.eventHubName,
        eventHubAuthorizationRuleId: data.eventHubAuthorizationRuleId,
        workspaceId: data.workspaceId,
        logAnalyticsDestinationType: data.logAnalyticsDestinationType,
        serviceBusRuleId: data.serviceBusRuleId,
        type: data.type,
        'log.Administrative': getLog(data.logs, 'Administrative')?.enabled,
        'log.Alert': getLog(data.logs, 'Alert')?.enabled,
        'log.Policy': getLog(data.logs, 'Policy')?.enabled,
        'log.Security': getLog(data.logs, 'Security')?.enabled,
      },
    },
  });
}

function getLog(logs: LogSettings[] | undefined, category: string) {
  if (!logs) return;

  const log = logs.find((l) => l.category === category);

  if (log) {
    return {
      enabled: log.enabled,
      retentionPolicyEnabled: log.retentionPolicy?.enabled,
      retentionPolicyDays: log.retentionPolicy?.days,
    };
  }
}
