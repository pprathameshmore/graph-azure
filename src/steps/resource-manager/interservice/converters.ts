import { VirtualMachine } from '@azure/arm-compute/esm/models';
import {
  NetworkInterface,
  PublicIPAddress,
  Subnet,
} from '@azure/arm-network/esm/models';
import {
  createDirectRelationship,
  Relationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { VIRTUAL_MACHINE_ENTITY_TYPE } from '../compute';
import { NetworkEntities } from '../network/constants';
import { VIRTUAL_MACHINE_NIC_RELATIONSHIP_TYPE } from './constants';

export function createVirtualMachineNetworkInterfaceRelationship(
  vm: VirtualMachine,
  nic: NetworkInterface,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.USES,
    fromKey: vm.id as string,
    fromType: VIRTUAL_MACHINE_ENTITY_TYPE,
    toKey: nic.id as string,
    toType: NetworkEntities.NETWORK_INTERFACE._type,
    properties: {
      _type: VIRTUAL_MACHINE_NIC_RELATIONSHIP_TYPE,
      vmId: vm.vmId as string,
    },
  });
}

export function createSubnetVirtualMachineRelationship(
  subnet: Subnet,
  vm: VirtualMachine,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromKey: subnet.id as string,
    fromType: NetworkEntities.SUBNET._type,
    toKey: vm.id as string,
    toType: VIRTUAL_MACHINE_ENTITY_TYPE,
  });
}

export function createVirtualMachinePublicIPAddressRelationship(
  vm: VirtualMachine,
  ipAddress: PublicIPAddress,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.USES,
    fromKey: vm.id as string,
    fromType: VIRTUAL_MACHINE_ENTITY_TYPE,
    toKey: ipAddress.id as string,
    toType: NetworkEntities.PUBLIC_IP_ADDRESS._type,
    properties: {
      vmId: vm.vmId as string,
    },
  });
}
