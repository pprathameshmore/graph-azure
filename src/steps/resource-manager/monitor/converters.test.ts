import { createAzureWebLinker } from '../../../azure';
import { createMonitorLogProfileEntity } from './converters';
import { LogProfileResource } from '@azure/arm-monitor/esm/models';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createMonitorLogProfileEntity', () => {
  test('properties transferred', () => {
    const data: LogProfileResource = {
      id:
        '/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/providers/microsoft.insights/logprofiles/default',
      type: '',
      name: 'default',
      location: '',
      storageAccountId:
        '/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/resourceGroups/JohnKemTest/providers/Microsoft.Storage/storageAccounts/johnkemtest8162',
      serviceBusRuleId: '',
      locations: ['global'],
      categories: ['Delete', 'Write', 'Action'],
      retentionPolicy: {
        enabled: true,
        days: 3,
      },
    };

    const logProfileEntity = createMonitorLogProfileEntity(webLinker, data);

    expect(logProfileEntity).toMatchSnapshot();
    expect(logProfileEntity).toMatchGraphObjectSchema({
      _class: ['Configuration'],
      schema: {
        additionalProperties: true,
        properties: {
          storageAccountId: { type: 'string' },
          serviceBusRuleId: { type: 'string' },
          locations: { type: 'array', items: { type: 'string' } },
          categories: { type: 'array', items: { type: 'string' } },
          'retentionPolicy.enabled': { type: 'boolean' },
          'retentionPolicy.days': { type: 'number' },
        },
      },
    });
  });
});
