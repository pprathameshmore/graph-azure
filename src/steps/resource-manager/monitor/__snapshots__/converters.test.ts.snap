// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMonitorLogProfileEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Configuration",
  ],
  "_key": "/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/providers/microsoft.insights/logprofiles/default",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "categories": Array [
          "Delete",
          "Write",
          "Action",
        ],
        "id": "/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/providers/microsoft.insights/logprofiles/default",
        "location": "",
        "locations": Array [
          "global",
        ],
        "name": "default",
        "retentionPolicy": Object {
          "days": 3,
          "enabled": true,
        },
        "serviceBusRuleId": "",
        "storageAccountId": "/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/resourceGroups/JohnKemTest/providers/Microsoft.Storage/storageAccounts/johnkemtest8162",
        "type": "",
      },
    },
  ],
  "_type": "azure_monitor_log_profile",
  "categories": Array [
    "Delete",
    "Write",
    "Action",
  ],
  "createdOn": undefined,
  "displayName": "default",
  "id": "/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/providers/microsoft.insights/logprofiles/default",
  "locations": Array [
    "global",
  ],
  "name": "default",
  "retentionPolicy.days": 3,
  "retentionPolicy.enabled": true,
  "serviceBusRuleId": "",
  "storageAccountId": "/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/resourceGroups/JohnKemTest/providers/Microsoft.Storage/storageAccounts/johnkemtest8162",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/df602c9c-7aa0-407d-a6fb-eb20c8bd1192/providers/microsoft.insights/logprofiles/default",
}
`;
